tools/python/m2ctx.py

```diff
@@ -1,71 +1,454 @@
+            for memType in structTypes:
+                if memType in types and memType not in nameStack:
+                    if memType != "Object": # This is a hack, due to circular referencing making things difficult. :/
+                        out += write_single_type(types, memType, addedTypes, nameStack)
+            out += types[name]['value'] + "\n"
+        elif kind == "enum":
+            out += types[name]['value'] + "\n"
+        addedTypes[name] = 1
+    nameStack.pop()
+    return out
+
+
+# Writes all the typedef/struct/union/enum to the output file
+def write_output_types(types):
+    out = ''
+    out += 'struct Object;\n' # This is a hack, due to circular referencing making things difficult. :/
+    addedTypes = {}
+    for name in types.keys():
+        nameStack = []
+        out += write_single_type(types, name, addedTypes, nameStack)
+    return out + '\n'
+
+# Writes all the function prototypes to the output file
+def write_output_functions(funcs):
+    out = ''
+    for name in funcs.keys():
+        out += funcs[name]['type'] + ' ' + name + '(' + funcs[name]['args'] + ');\n'
+    return out + '\n'
+
+# Writes all the preprocessor directives to the output file
+def write_output_directives(directivesArray):
+    out = ''
+    out += '#define NULL (void*)0\n' # Hack, since cleanup() will remove the original NULL definition.
+    for direct in directivesArray:
+        out += direct + '\n'
+    return out + '\n'
+
+# Writes a variable to the output file
+def write_output_variable(key, variables, varKeys, doneKeys):
+    if key in doneKeys:
+        return ""
+    out = ""
+    text = variables[key]
+    for key2 in varKeys:
+        if len(variables[key2]) < 1:
+            continue
+        if key == key2:
+            continue
+        if key2 in text:
+            out += write_output_variable(key2, variables, varKeys, doneKeys)
+    out += text + '\n'
+    doneKeys[key] = 1
+    return out
+
+# Writes all the variables to the output file
+def write_output_variables(variables):
+    out = ''
+    doneKeys = {}
+    varKeys = variables.keys()
+    for key in varKeys:
+        if len(variables[key]) < 1:
+            continue
+        out += write_output_variable(key, variables, varKeys, doneKeys)
+    return out + '\n'
+
+# Creates the output file text.
+def write_output(data):
+    out = '/*** DKR decomp context file (Automatically generated by m2ctx.py [' + date.today().strftime("%Y/%m/%d") + ']) ***/\n\n'
+    out += write_output_directives(data['directives'])
+    out += write_output_types(data['types'])
+    out += write_output_variables(data['variables'])
+    out += write_output_functions(data['functions'])
+    out = fix_enums(out) # Fixes trailing commas in enums, which gets rid of a warning.
+    out = cleanup(out) # Removes useless stuff
+    return out
+    
+# Uses the `find` program to return the filenames in the specified folders.
+def find_files():
+    find_command = ["find"] + search_folders
```

Comment:
Previously, shell script was using `find <DIRS> -type f -name "*.h"` - should the python way also only header files?

Reply:
I added the .c files to get the variables. If we were to only do headers we would have to have an extern for every single variable in the header files.

Reply:
My concern is also that find will pick up directories in addition to files and could potentially also pick up non-C/H files. May I recommend the following?
```sh
find <search_folders> -type f -name "*.[ch]"
```

Reply:
Ah nice. I didn't know how to get both .c & .h files using the find command, tried a bunch of different patterns but they didn't work. I just decided to get every file and sort it in the python script using `os.path.isfile(filename)`. I'll add your suggestion though.